### PR TITLE
Add vSphere provider support

### PR DESCRIPTION
**Pull request by luomiao**
This PR enables Kops basic functionality over vSphere platform.
Currently the support for vSphere is behind feature flag VSphereCloudProvider.

General guideline about how to use kops over vSphere can be found at: docs/development/vsphere-dev.md

There are some temporary work-arounds in this PR:
1. Due to the gap between upstreaming and release, there are changes to enable user to define different binary/image for nodeup/protokube/dns-controller. These changes can be removed after the vSphere supported is fully included into release binaries.
2. vSphere provider by default uses CoreDNS as DNS service provider. Since one of the interface functions used by current dns-controller in kops (ResourceRecordSets.List()) is not implemented in CoreDNS provider yet, there are work-arounds added inside dns-controller related code. Those work-arounds only affect when the DNS provider is set to CoreDNS. Those work-arounds can be removed after the missing interface function is implemented in CoreDNS dns provider.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2401)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Can we get a squash?

I have reviewing this on my radar, but I need to have some time.

Would love to see it in action.

**Comment by luomiao** [comment]
@chrislovecnm Hi Chris, do you suggest a squash now for review or do you mean the squash when merge?

File: protokube/cmd/protokube/main.go [hunk 1]
```diff
@@ -126,6 +132,18 @@ func run() error {
...
+	} else if cloud == "vsphere" {
+		glog.Info("Initializing vSphere volumes")
+		vsphereVolumes, err := protokube.NewVSphereVolumes()
+		if err != nil {
+			glog.Errorf("Error initializing vSphere: %q", err)
+			os.Exit(1)
+		}
+		volumes = vsphereVolumes
+		if internalIP == nil {
+			internalIP = vsphereVolumes.InternalIp()
```
Comment (justinsb) [inline protokube/cmd/protokube/main.go]: Nit: golang style says we should use `InternalIP`.  But I think it looks wrong too :-)

File: protokube/pkg/protokube/vsphere_volume.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		lines := strings.Split(string(out), "\n")
+		for _, line := range lines {
+			if strings.Contains(line, mountPoint) {
+				lsblkOutput := strings.Split(line, " ")
+				glog.V(4).Infof("Found device: %v ", lsblkOutput[0])
+				return "/dev/" + lsblkOutput[0], nil
+			}
+		}
+	} else {
+		return "", fmt.Errorf("Failed to find device. OS %v is not supported for vSphere.", runtime.GOOS)
```
Comment (justinsb) [inline protokube/pkg/protokube/vsphere_volume.go]: Well this is interesting :-)

**Comment by justinsb** [comment]
I haven't reviewed the vmware specific code in much detail, because it is feature flagged and I figure you know best, but this LGTM :-)

**Pull request merged**